Ios 5738 algorand make history txs

skibinalexander opened the pull request:
Рефакторинг истории транзакции для отображения истории в Algorand. Рефакторинг требовался из-за того, что история транзакций в Algorand не по страничная, а связанным списком. Пока смерджим #533 будет WIP

skibinalexander committed:
IOS-5738 Merge branch 'IOS-5588-make-algorand-network-layer' into IOS-5738-algorand-make-history-txs

skibinalexander committed:
IOS-5738 Make basic implementation history provider

skibinalexander committed:
IOS-5738 Fix algorand history changes

skibinalexander committed:
IOS-5738 Merge branch 'IOS-5588-make-algorand-network-layer' into IOS-5738-algorand-make-history-txs

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj

skibinalexander committed:
IOS-5738 Fix by merge

skibinalexander committed:
IOS-5738 fix error

skibinalexander committed:
IOS-5738 make history transaction

skibinalexander committed:
IOS-5738 fix changes

skibinalexander committed:
IOS-5738 IOS-5723 Make transaction history algorand domain

skibinalexander committed:
IOS-5738 Merge branch 'IOS-5591-algorand-integration' into IOS-5738-algorand-make-history-txs

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/BlockchainService/Algorand/Provider/AlgorandProviderTarget.swift
#	BlockchainSdk/BlockchainService/Algorand/Provider/AlgorandProviderUrl.swift

skibinalexander committed:
IOS-5738 Merge branch 'IOS-5591-algorand-integration' into IOS-5738-algorand-make-history-txs

skibinalexander committed:
IOS-5738 Merge branch 'develop' into IOS-5738-algorand-make-history-txs

skibinalexander committed:
IOS-5738 Merge branch 'develop' into IOS-5738-algorand-make-history-txs

skibinalexander committed:
IOS-5738 Merge branch 'develop' into IOS-5738-algorand-make-history-txs

Balashov152 commented on BlockchainSdk/BlockchainService/Algorand/DTO/AlgorandResponse+TransactionHistory.swift:
```diff
@@ -0,0 +1,91 @@
+            transactions = try container.decode([Item].self, forKey: .transactions)
+        }
+
+        private enum CodingKeys: String, CodingKey {
+            case nextToken = "next-token"
+            case currentRound = "current-round"
+            case transactions
+        }
```
Поискал `convertFromKebabCase` но там кода примерно столько же как и здесь)) Так что лучше так наверно оставить все

skibinalexander replied:
> Поискал `convertFromKebabCase` но там кода примерно столько же как и здесь)) Так что лучше так наверно оставить все

Может чуть позже рефакторингом сделаю

Balashov152 commented on BlockchainSdk/BlockchainService/Algorand/Provider/AlgorandProviderTarget.swift:
```diff
@@ -54,6 +58,17 @@ struct AlgorandProviderTarget: TargetType {
         case .transaction(let data):
             return .requestData(data)
+        case .getTransactions(let address, let limit, let next):
+            let parameters: [String: Any?] = [
+                "address": address,
+                "limit": limit,
+                "next": next
+            ]
```
Я бы добавил 

```
enum AlgorandTransactionAPI {}
extension AlgorandTransactionAPI  { 
    struct  Response { ... } 
    struct  Request { ... } 
}
 ```
 
 и вот здесь бы AlgorandTransactionAPI. Request использовал, и моделька сам `Encodable` бы делала

skibinalexander replied:
Переделал

Balashov152 commented on BlockchainSdk/Common/TransactionHistory/Algorand/AlgorandTransactionHistoryMapper.swift:
```diff
@@ -0,0 +1,46 @@
+                    .init(address: .user($0.paymentTransaction.receiver), amount: decimalAmountValue)
+                ),
+                fee: .init(.init(with: blockchain, value: decimalFeeValue)),
+                status: .confirmed,
```
А если не `.confirmed` будет? или таких не бывает транзакций в истории ?

skibinalexander replied:
> А если не `.confirmed` будет? или таких не бывает транзакций в истории ?

сейчас еще раз проверю

skibinalexander replied:
> А если не `.confirmed` будет? или таких не бывает транзакций в истории ?

https://developer.algorand.org/docs/rest-apis/indexer/#get-v2transactions

По идее вываливает только подтвержденные транзакции

Balashov152 replied:
лишь бы не терялось для пользователя )

Balashov152 commented on BlockchainSdk/Common/TransactionHistory/TransactionHistory/Ethereum/EthereumTransactionHistoryProvider.swift:
```diff
@@ -25,10 +29,31 @@ class EthereumTransactionHistoryProvider {
+        page == nil || (page?.number ?? 0) < totalPages
+    }
+    
+    var debugDescription: String {
```
Наверно надо добавить `TransactionHistoryProvider: CustomStringConvertable` что бы это норм работало

skibinalexander replied:
> TransactionHistoryProvider

Тут не уверен, но запрос был от @tureck1y так сделать, более гибче. 
@tureck1y подскажи правильно тебя понял? оставляем?

Balashov152 replied:
Это штука идет для `print()` стандартного. 
Если объект конформит, то распечатается как раз эта стринга.

`CustomDebugStringConvertible -> debugDescription`
`CustomStringConvertible  -> description`

там простой протокол, если не конфирмит, то что то свое там swift подложит

Balashov152 commented on BlockchainSdk/Common/TransactionHistory/Page.swift:
```diff
@@ -8,12 +8,12 @@
-        self.size = size
-    }
+// Use for indexed navigation state
+struct TransactionHistoryIndexPage: Hashable {
```
Файл наверно тоже надо переименовать

Balashov152 commented on BlockchainSdk/Common/TransactionHistory/TransactionHistoryProvider.swift:
```diff
@@ -11,5 +11,9 @@ import Combine
 @available(iOS 13.0, *)
 public protocol TransactionHistoryProvider {
+    var canFetchHistory: Bool { get }
+    var debugDescription: String { get }
```
Да, и вот здесь убрать можно, если `CustomDebugStringConvertible` или `CustomStringConvertible` 
он сам через print будет работать, когда в App вызовешь

skibinalexander replied:
> Да, и вот здесь убрать можно, если `CustomDebugStringConvertible` или `CustomStringConvertible` он сам через print будет работать, когда в App вызовешь

Коммент выше

skibinalexander committed:
IOS-5738 Merge branch 'develop' into IOS-5738-algorand-make-history-txs

# Conflicts:
#	BlockchainSdk/BlockchainService/Algorand/AlgorandNetworkService.swift

skibinalexander committed:
IOS-5738 Fix review comments

skibinalexander committed:
IOS-5738 Add CustomStringConvertible for transaction history provider

skibinalexander committed:
IOS-5738 Replace debugDescription to description
